src/test/storage.zig

```diff
@@ -149,41 +179,93 @@ pub const Storage = struct {
         try writes.ensureTotalCapacity(config.io_depth_write);
 
         return Storage{
+            .allocator = allocator,
             .memory = memory,
-            .memory_occupied = memory_occupied,
-            .size = size,
+            .memory_written = memory_written,
             .faults = faults,
+            .size = size,
             .options = options,
-            .replica_index = replica_index,
             .prng = std.rand.DefaultPrng.init(options.seed),
-            .faulty_areas = faulty_areas,
             .reads = reads,
             .writes = writes,
         };
     }
 
-    /// Cancel any currently in progress reads/writes but leave the stored data untouched.
+    /// Cancel any currently in-progress reads/writes.
+    /// Corrupt the target sectors of any in-progress writes.
     pub fn reset(storage: *Storage) void {
-        while (storage.writes.peek()) |write| {
-            _ = storage.writes.remove();
-            // TODO Allow some corruption in other zones.
-            if (write.zone == .wal) {
-                storage.fault_sectors(write.offset, write.buffer.len);
-            }
+        while (storage.writes.peek()) |_| {
+            const write = storage.writes.remove();
+            if (switch (write.zone) {
+                .superblock => !storage.options.faulty_superblock,
+                // On crash, the WAL may be corrupted outside of the FaultyAreas.
+                .wal_headers, .wal_prepares => storage.options.faulty_wal_areas == null,
+                // TODO Enable fault injection for grid.
+                .grid => true,
+            }) continue;
+
+            if (!storage.x_in_100(storage.options.crash_fault_probability)) continue;
+
+            const sector_min = @divExact(write.zone.offset(write.offset), config.sector_size);
+            const sector_max = sector_min + @divExact(write.buffer.len, config.sector_size);
+
+            // Randomly corrupt one of the faulty sectors the operation targeted.
+            // TODO: inject more realistic and varied storage faults as described above.
+            storage.fault_sector(
+                write.zone,
+                storage.random_uint_between(usize, sector_min, sector_max),
+            );
         }
+        assert(storage.writes.len == 0);
 
         storage.reads.len = 0;
-        assert(storage.writes.len == 0);
     }
 
     pub fn deinit(storage: *Storage, allocator: mem.Allocator) void {
         allocator.free(storage.memory);
-        storage.memory_occupied.deinit(allocator);
+        storage.memory_written.deinit(allocator);
         storage.faults.deinit(allocator);
         storage.reads.deinit();
         storage.writes.deinit();
     }
 
+    pub fn size_used(storage: *const Storage) usize {
+        var sector_iterator = storage.memory_written.iterator(.{ .direction = .reverse });
+        const sector_max = sector_iterator.next() orelse 0;
+        return (sector_max + 1) * config.sector_size;
+    }
+
```

Comment:
I was using memory_written.count() instead. I guess this calculation is assuming that unwritten sectors before the last last written sector still occupy storage, which leads me to realize that I don't know how file-systems work.  If I open a file and write to sector 1000000, do the preceding 999999 sectors get created on disk?

Reply:
It depends on the file system. I think most of the time, the file would be sparse, and only the last sector would be written to disk. i.e. The underlying sectors referenced by a file are not guaranteed to be contiguous. How this impacts the filesystems view of available storage on disk might be a separate matter.

OSTEP is great (along with any storage paper from the Arpaci-Dusseau's): https://pages.cs.wisc.edu/~remzi/OSTEP/

Reply:
Perfect, I'll change this to `.count()` then. Thanks @jamii & @jorangreef!
(Also, how did I not know that OSTEP was written by the Arpaci-Dusseau's?!)

Reply:
Thanks @sentientwaffle for the ping!

Re: `fallocate()` that would allocate the sectors, and ensure that they are zeroed. But there are optimizations in how it does this: https://man7.org/linux/man-pages/man2/fallocate.2.html